services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushStatusCollector.java

```diff
@@ -0,0 +1,231 @@
+      try {
+        pushStatus = future.get();
+      } catch (Exception e) {
+        LOGGER.error("Caught exception when getting future result of push status.", e);
+        continue;
+      }
+      ExecutionStatusWithDetails daVinciStatus = pushStatus.getDaVinciStatus();
+      ExecutionStatusWithDetails serverStatus = pushStatus.getServerStatus();
+      if (serverStatus == null) {
+        continue;
+      }
+      LOGGER.info(
+          "Topic server push status: {}, Da Vinci push status: {}",
+          serverStatus.getStatus(),
+          daVinciStatus.getStatus());
+      if (isOverallPushCompleted(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        pushCompletedHandler.accept(pushStatus.getTopicName());
+      } else if (isOverallPushError(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        StringBuilder pushErrorDetailStringBuilder = new StringBuilder();
+        if (serverStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          pushErrorDetailStringBuilder.append("Server push error: ").append(serverStatus.getDetails()).append("\n");
+        }
+        if (daVinciStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          pushErrorDetailStringBuilder.append("Da Vinci push error: ").append(daVinciStatus.getDetails()).append("\n");
+        }
+        pushErrorHandler.accept(pushStatus.getTopicName(), pushErrorDetailStringBuilder.toString());
+      }
+    }
+  }
+
+  public void handleServerPushStatusUpdate(String topicName, ExecutionStatus executionStatus, String detailsString) {
+
+    TopicPushStatus topicPushStatus = topicToPushStatusMap.get(topicName);
+    if ((!daVinciPushStatusScanEnabled) || topicPushStatus == null) {
+      if (executionStatus.equals(ExecutionStatus.COMPLETED)) {
+        pushCompletedHandler.accept(topicName);
+      } else if (executionStatus.equals(ExecutionStatus.ERROR)) {
+        pushErrorHandler.accept(topicName, detailsString);
+      }
+    } else {
+      // Update the server topic status in the data structure and wait for async DVC status scan thread to pick up.
+      topicToPushStatusMap.computeIfPresent(topicName, (topic, pushStatus) -> {
+        pushStatus.setServerStatus(new ExecutionStatusWithDetails(executionStatus, detailsString));
+        return pushStatus;
+      });
+    }
+  }
+
+  private boolean isOverallPushCompleted(
+      ExecutionStatusWithDetails serverStatus,
+      ExecutionStatusWithDetails daVinciStatus) {
+    return serverStatus.getStatus().equals(ExecutionStatus.COMPLETED)
+        && daVinciStatus.getStatus().equals(ExecutionStatus.COMPLETED);
+  }
+
+  private boolean isOverallPushError(
+      ExecutionStatusWithDetails serverStatus,
+      ExecutionStatusWithDetails daVinciStatus) {
+    return serverStatus.getStatus().equals(ExecutionStatus.ERROR)
+        || daVinciStatus.getStatus().equals(ExecutionStatus.ERROR);
+  }
+
+  public void clear() {
+    topicToPushStatusMap.clear();
```

Comment:
Do we need to shutdown `pushStatusStoreScanExecutor` and `offlinePushCheckScheduler` here?

Reply:
Looks like we cannot do it explicitly here. The other PushMonitor does not implement VeniceService interface and does not have place to stop() explicitly. The place to call clear() is stopMonitoringAllPushes(), where it is not stopping the service and could actually be called during controller state transition.
So far I think it is not easy to do that, but maybe we could refactor it in a separate PR I guess?

Reply:
Actually, let me double check whether it is easy to add some stop() like thing to PushMonitor